### ISSUE TITLE
Run Jackett on Tomato router (Busybox 1.25 + nginx)

### ISSUE THREAD

**Issue by wellloaded** [Jackett/Jackett#6615]
Hi there

I can see you have a release of Jackett for ARMv7 32bit which is great.
I have Tomato running on my router over ARMv7 Hardware (Kernel 2.6)

The Tomato Firmware comes with a built in nginx webserver however I only have busybox 1.25 available which comes with built in /bin/sh

```
BusyBox v1.25.1 (2019-11-19 23:02:03 CET) multi-call binary.
BusyBox is copyrighted by many authors between 1998-2015.
Licensed under GPLv2. See source distribution for detailed
copyright notices.

Usage: busybox [function [arguments]...]
   or: busybox --list
   or: function [arguments]...

        BusyBox is a multi-call binary that combines many common Unix
        utilities into a single executable.  Most people will create a
        link to busybox for each function they wish to use and BusyBox
        will act like whatever it was invoked as.

Currently defined functions:
        [, [[, arp, arping, ash, awk, basename, blkid, cat, chmod, chown, chroot, clear, cmp, cp, crond, cut, date, dd, df, dirname, dmesg, dnsdomainname, du, e2label, echo, egrep, env, ether-wake, expr, fdisk, fgrep, find, flock, free, fsync, ftpget, ftpput, grep, gunzip,
        gzip, head, hostname, ifconfig, insmod, ionice, kill, killall, klogd, less, ln, logger, login, ls, lsmod, lsusb, md5sum, mkdir, mkdosfs, mkfifo, mkfs.vfat, mknod, mkswap, modprobe, more, mount, mv, nc, netstat, nice, nohup, nslookup, ntpd, pidof, ping, ping6, printf,
        ps, pscan, pwd, rm, rmdir, rmmod, route, sed, sendmail, seq, setconsole, sh, sleep, sort, strings, stty, swapoff, swapon, sync, syslogd, tail, tar, tee, telnet, telnetd, test, top, touch, tr, traceroute, traceroute6, true, udhcpc, umount, uname, unzip, uptime, usleep,
        vconfig, vi, watch, wc, wget, which, whois, zcat

```

Is there any ways to have Jackett running directly on my router?

Thanks!!

**Comment by garfield69** [comment]
@MichaIng  @ngosang you both have far more experience with Linux platforms than I.
Can you answer the OP's question in this ticket?
My first reaction is that if the OP's particular flavour of Linux does not provide the standard tools, then its likely that the ARM32 package will not install without the OP having to manually shoehorn it with changes to the scripts.
But there will probably be ongoing issues with the updater and whatnot. Not something we can get involved in supporting long term.

**Comment by ngosang** [comment]
@wellloaded I think it should work. Try this:
Download the latest version for arm 32 => Jackett.Binaries.LinuxARM32.tar.gz
Uncompress with tar command
cd Jackett
run ./jackett

**Comment by wellloaded** [comment]
> 
> 
> @wellloaded I think it should work. Try this:
> Download the latest version for arm 32 => Jackett.Binaries.LinuxARM32.tar.gz
> Uncompress with tar command
> cd Jackett
> run ./jackett

router# ./jackett
-sh: ./jackett: not found

As I read there's a set of libraries that need to be made available to Jackett
https://github.com/dotnet/core/blob/master/Documentation/linux-prereqs.md

**Comment by MichaIng** [comment]
@wellloaded 
> As I read there's a set of libraries that need to be made available to Jackett

Here is another list: https://docs.microsoft.com/de-de/dotnet/core/install/dependencies?pivots=os-linux&tabs=netcore30#linux-distribution-dependencies
Most are installed by default on Linux distros, but in your case at least it is worth to verify that.

> router# ./jackett
-sh: ./jackett: not found

This simply looks like you're in the wrong directory. The binary is inside the "Jackett" directory that is contained in the archive, hence the `cd Jackett` mentioned by ngosang. Or run `Jackett/jackett` after extracting the archive.

Since the .NET standalone binary is quite self-contained and portable, it should otherwise run fine on the Tomato, regardless of the shell tools capability. Of course the systemd and very likely as well sysvinit services will not work, but running the binary from console should not be affected. This should include the updater as well.

**Comment by wellloaded** [comment]
> This simply looks like you're in the wrong directory. The binary is inside the "Jackett" directory that is contained in the archive, hence the "`cd Jackett`" mentioned by ngosang. Or run `Jackett/jackett` after extracting the archive.
```

root@tomato20n:/tmp/home/root# cd /mnt/usb/nginx/html/Jackett/
root@tomato20n:/tmp/mnt/usb/nginx/html/Jackett# pwd
/mnt/usb/nginx/html/Jackett
root@tomato20n:/tmp/mnt/usb/nginx/html/Jackett# ls -la jackett*
-rwxr-xr-x    1 root     root         73400 Dec 15 07:12 jackett
-rwxr-xr-x    1 root     root        258999 Dec 15 07:12 jackett.deps.json
-rwxr-xr-x    1 root     root        392704 Dec 15 07:12 jackett.dll
-rwxr-xr-x    1 root     root         20428 Dec 15 07:12 jackett.pdb
-rwxr-xr-x    1 root     root            93 Dec 15 07:12 jackett.runtimeconfig.json
-rwxr-xr-x    1 root     root           402 Dec 15 07:11 jackett_launcher.sh
root@tomato20n:/tmp/mnt/usb/nginx/html/Jackett# ./jackett
-sh: ./jackett: not found

```

**Comment by garfield69** [comment]
I am not surprised. I really cant see how you can run applications on firmware software made to run and control routers. I do not think this OS is a general applications platform.
Are you able to run a browser on this platform, or run any other general application?
If so, how do you launch them?

**Comment by wellloaded** [comment]
I personally use [Freshtomato](www.freshtomato.org) but there are many flavours out there. You might compare Tomato to DD-WRT or OPENWRT they are similar projects.

So yes let me explain. [Tomato](https://en.wikipedia.org/wiki/Tomato_(firmware)) is based on Linux 2.6. It's a squashfs so the core system is coming from Flash at every boot and is loaded into memory so you can't modify that permanently unless you recompile the sources.

However, since it has been developed for many years now it has many options available, the most important is USB/CIFS/JFFS support where statically compiled applications can be stored permanently and run within the Linux environment.

There are actually a couple of officially maintained repositories of "ready to run" application (Optware/Entware) where you can fetch a binary store it on local non volatile storage and run them.

It is actually a surprisingly good system to be running on a router and does a huge amount of things like webserver/mysql/media server/VPN/FTP/TFTP/bittorrent client/and much, much more.

The idea would be to run Jackett on the router and access it from the LAN clients, where the download can be handled either by a torrent client installed on the end client or on the bittorrent client available on the router itself (Transmission)

You run applications the way you would do on a full Linux. Exactly the same. Somebody made PLEX running on Tomato, at the end of the day it's Linux so it's just that applications that need to be compiled in a certain way to run.

**Comment by MichaIng** [comment]
@wellloaded 
Your error definitely states that the file has not been found:
```
# ./doesnotexist
sh: 1: ./doesnotexist: not found
```
Probably your shell simply does not understand the `.` to be the current dir, hence `./exec` fails. I'm no expert but the `.` and `..` are AFAIK "real" items being created, not a special syntax, so probably those are missing on either R/O or squashfs in particular. Does "`ls . ..`" show the content of current and parent dir correctly?

However full path should then work:
```
/mnt/usb/nginx/html/Jackett/jackett
```

@garfield69 
I don't think this being a fundamental issue. Tomato is still Linux and one can configure it and install anything required offline, hence it must be able to work, as far as Jackett does not fail when expecting FS write capability at some point. The shell, whether being bash, dash, zsh or busybox should not make any difference, as long as we find the correct syntax to execute a binary, when `./exec` does not work 😄.

**Comment by wellloaded** [comment]
> However full path should then work:
> 
> ```
> /mnt/usb/nginx/html/Jackett/jackett
> ```

AFAIK this is normal behaviour for executable not compatible with the HW or incorrect EOL for scripts:
```

root@tomato20n:/tmp/mnt/usb/nginx/html/Jackett# ls -la /mnt/usb/nginx/html/Jackett/jackett
-rwxr-xr-x    1 root     root         73400 Dec 15 07:12 /mnt/usb/nginx/html/Jackett/jackett
root@tomato20n:/tmp/mnt/usb/nginx/html/Jackett# /mnt/usb/nginx/html/Jackett/jackett
-sh: /mnt/usb/nginx/html/Jackett/jackett: not found

```

**Comment by MichaIng** [comment]
@wellloaded 
Okay, I get other error messages when binary has wrong architecture or script contains syntax error. However you will know best how to execute executables on Tomato shell/busybox. Did you assure that the above mentioned .NET dependency libraries are made available via some shared library default path and/or ldconfig or similar mechanism? The exact naming or version number can very, however actually I would expect some moreless clear error message from binary, when some library is missing, not "not found" 😉.

**Comment by ngosang** [comment]
> I am thinking if Linux 2.6 is an issue here

It's not a problem. Tested.

**Comment by avengermsoft** [comment]
@MichaIng @wellloaded  See https://github.com/Jackett/Jackett/pull/12886#issuecomment-1046033860

**Comment by garfield69** [comment]
Closed as stale.

**Comment by HommeOursPorc** [comment]
I've noticed that ash on FreshTomato tends to output a "not found" error when it encounter a Windows line return "\r". I tried to run the ./jackett script after removing the "\r", it now give a syntax error on line 1. It look like it is looking for some library that is obviously missing in FreshTomato. Also I suppose the jackett scripts are meant to be run on bash rather than busybody ash.

@wellloaded maybe it could run using some Entware stuffs?
